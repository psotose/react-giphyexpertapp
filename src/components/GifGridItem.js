import React from 'react'
import PropTypes from 'prop-types';

const GifGridItem = ( {title, url} ) => {  // se pone props en vez de img xq es lo que es en relidad ( props )
                                          // pero si desestructuro es mas facil de leer
  
  // console.log({id, title, url});
  return (
    <div className="card animate__animated animate__bounce">
      <img src={url} alt={title} />
      <p> { title } </p>  
      {/* { img.title } */}
    </div>
  )
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}
export default GifGridItem;
