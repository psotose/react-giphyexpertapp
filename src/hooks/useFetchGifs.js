// con el hook a diferencia del functional component, no necesito
// hacer el import de react a menos que retornemos JSX

import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs"

export const useFetchGifs = ( category ) => {
  const [state, setState] = useState({
    data: [],
    loading: true,
  });

  useEffect( () => {
    getGifs( category )
      .then( imgs =>{
        console.log(imgs)
        setState({
          data: imgs,
          loading: false
        });
      })
  }, [category])

  // setTimeout(() =>{
  //   setState({
  //     data: [1,2,3,4,5],
  //     loading: false
  //   })
  // }, 3000);

  return state; // { data:[], loading: true };
}