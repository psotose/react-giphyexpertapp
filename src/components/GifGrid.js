import React from 'react' //, { useState, useEffect } le saco esto xq voy a trabajar en el custom hook
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';


export const GifGrid = ({ category }) => {
  
  // const [count, setCount] = useState(0);
  // const state = useFetchGifs();
  const { data:images, loading} = useFetchGifs( category );
  // con data:images como quien dice renombro a data y luego puedo llamarlo como images
  //const [images, setImages] = useState([]);

  // useEffect( () => {
  //   getGifs( category )
  //     .then( imgs => setImages( imgs )); // esto es lo mismo que .then(setImages)
  // }, [ category ]) // con el [] le estoy diciendo que solo quiero que se ejecute la instrucci√≥n cuando el componente es renderizado por 1ra vez

  return (
    <>
      <h3 className="animate__bounceInRight"> {category} </h3>
      {/* { loading ? 'Cargando...' : 'Fin de carga' } */}
      { loading && <p className="animate_animated animate_flash">'Cargando...'</p> }  
      {/* con el && solo evalua el 1ro si es true y con el false no hace nada */}
      <div className="card-grid">
        {
          images.map( img => (  //antes del custom hook era images.map, despues de la asignacion data:images puedo volver a usar images
            <GifGridItem 
              key={ img.id }
              {...img}
              // img={ img } 
            />
          ))
        }     

        {/* <ol>
          {
            images.map( img => (
              <li key={ img.id }> { img.title } </li>
            ))
          }
          ####################################
          aca tb puedo desestructurar y hacer 
            images.map( { id, title } => (
              <li key={ 
                id }> { title } </li>
            )) 
          ######################################  
        </ol> 
           <h3> { count } </h3>
        <button onClick={ ()=> setCount( count + 1 ) }></button>*/}
      </div> 
    </>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
}

