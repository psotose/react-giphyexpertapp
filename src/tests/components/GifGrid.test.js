import react from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe('Testing GifGrid', () => {
  const category = 'alo';
  
  test('should be shown correctly', () => {
    useFetchGifs.mockReturnValue({ 
      data: [], 
      loading: true
    });
    const wrapper = shallow(<GifGrid category={ category } />);
    expect ( wrapper ).toMatchSnapshot();
  });

  test('should show items when useFetchGifs images are loaded', () => {

    const gifs = [{
      id: 'ABC',
      url: 'https://example.jpg',
      title: 'common'
    }];

    useFetchGifs.mockReturnValue({ 
      data: gifs, 
      loading: false
    });
    const wrapper = shallow(<GifGrid category={ category } />);

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find('p').exists() ).toBe(false);
    expect ( wrapper.find('GifGridItem').length ).toBe( gifs.length );
  });

});