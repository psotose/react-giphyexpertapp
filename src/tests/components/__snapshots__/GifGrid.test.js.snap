// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifGrid should be shown correctly 1`] = `
<Fragment>
  <h3
    className="animate__bounceInRight"
  >
     
    alo
     
  </h3>
  <p
    className="animate_animated animate_flash"
  >
    'Cargando...'
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Testing GifGrid should show items when useFetchGifs images are loaded 1`] = `
<Fragment>
  <h3
    className="animate__bounceInRight"
  >
     
    alo
     
  </h3>
  <div
    className="card-grid"
  >
    <Component
      id="ABC"
      key="ABC"
      title="common"
      url="https://example.jpg"
    />
  </div>
</Fragment>
`;
